src/middlewares.rs

```diff
@@ -6,29 +6,49 @@ use axum::{
 };
 use axum_extra::extract::CookieJar;
 
-#[derive(Clone, Debug)]
+use serde::{Deserialize, Serialize};
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct Course {
+    pub name: String,
+    pub crns: Vec<String>,
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
 pub struct UserState {
-    pub course_reg_numbers: Vec<String>,
+    pub selection: Vec<Course>,
 }
 
-pub type CookieUserState = Extension<Option<UserState>>;
+pub type CookieUserState = Extension<UserState>;
```

Comment:
by making CookieUserState not an option, it is required that we parse it in middleware and do error-handling at the source. that way, by the time we get to our request handlers, we can **guarantee** that the userstate it valid (either the `parse_cookie` function created a new state (cookie wasn't present), parsed it correctly (cookie was valid), or rejected the response/closed the request/reset the sate (cookie was malformed)

this simplifies our route handlers a bit because they no longer need to check for the validity of the user state

Reply:
This change makes sense to me since an empty state is possible. We should  just implement `Default` on `UserState` if we do this.

On a related note, it might be worth to build this out as a [custom Extractor instead](https://docs.rs/axum/latest/axum/extract/index.html#defining-custom-extractors), since that's basically what the middleware is doing right now as I understand it. We could try forwarding the `parts` and `state` values to `CookieJar` and then operate on the results.

How do you feel about this @hn275? I'm curious if there are any aspects of the design that I'm totally not seeing here.

Reply:
this is merged, but I agree on all points you guys made so far. 

> I'm curious if there are any aspects of the design that I'm totally not seeing here.

there wasn't any design aspect, I was just trying to get an extractor going and my own expectation of control flow, but handling errors at the source does sound to be the smarter move.

> We should just implement Default on UserState if we do this.
>
> On a related note, it might be worth to build this out as a [custom Extractor instead](https://docs.rs/axum/latest/axum/extract/index.html#defining-custom-extractors)

@brennanmcmicking i like this idea, I'll get a pr up for this.